docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
+If the user does not have the dynamic privileges granted in MySQL, the overall approval will fail and the user will get a privilege error. The overall logic should be:
+```go
+func VerifyDynamicPrivilege(mysqlPriv *MySQLPrivilege, user string) bool {
+	return mysqlPriv.RequestDynamicVerification(...) && getPluginForUser(user).RequestDynamicVerification(...)
+}
+```
+
+Note that for both the privilege checking functions above, they only receive one privilege in the parameter. This means if the transactions include multiple statements requiring multiple privileges, these functions will be called multiple times using the different privileges.
+
+The above interfaces and parameters should give the users all information they need to make a decision on the authentication and authorization. For authorizations before each request, the MySQL user should already be granted with the corresponding privileges in MySQL using `GRANT` statements.
+
+### Compatibility
+The feature needs to be compatible with existing user related statements and options, such as `CREATE USER`, `ALTER USER`, `SET PASSWORD`.
+However, it will not work with existing functionalities such as password expiry and password validation. Users can implement their own password validation logic inside the EncodePassword implementation.
+
+## Test Design
+
+This feature requires tests focusing on functional and compatibility. Performance should not be a concern, since that depends on the actual auth plugin implementation and it is up to the users to keep the implementation performant if needed.
+
+### Functional Tests
+
+We will implement unit tests for all of the auth plugin extension interfaces, and ensure that they are executed properly in the code path that we inject them.
+
+### Scenario Tests
+
+We will implement client to server tests to test out scenarios where MySQL users are created with and without using an auth plugin from the extension system. For example:
++ Creating a user using the plugin
+  + Connecting as the user and verify that the connection verification API is called and errors are correctly translated to connection rejections
+    + Execute SQL statements and verify that the request (dynamic) verification APIs are called and errors are correctly translated to permission rejections
+    + Alter user changes can be applied correctly
+    + Passwords can be encoded with the API defined in the plugin
++ When the connection changes between users, the system should behave correctly
+  + From non-plugin user to plugin user
+  + From plugin user to non-plugin user
+  + From plugin user to plugin user
+
+### Compatibility Tests
+
+This feature should integrate well with all the existing test cases since it is an additional functionality. Existing tests related to authentication and authorization should continue to pass since they are not using auth plugins.
```

Comment:
Maybe we need to define the behavior of some other command like `SHOW GRANTS FOR <someuser>` when "someuser" uses the custom plugin.

Reply:
Or the view behavior when a view's definer is a user with a custom plugin. (Whether or not to check a definer's privilege declared in the plugin).

Reply:
> define the behavior of some other command like `SHOW GRANTS FOR <someuser>` when "someuser" uses the custom plugin.

I intend to keep the behavior the same as the current implementation, where `SHOW GRANTS` will only display existing grants based on internal MySQL privilege table. The reason is that if `VerifyPrivilege` depends on an outside system, then the privileges granted might require other information such as the SSL certificates, which is not available as a static information when fetching the MySQL grants. Please let me know what do you think.

> view behavior when a view's definer is a user with a custom plugin. (Whether or not to check a definer's privilege declared in the plugin).

I'm not too sure if we need to change anything for this. IIUC view definer's privilege is only checked when running `CREATE VIEW`, which the plugin will need to call `VerifyPrivilege` to check if the definer has `CREATE VIEW` priv. Other operations for the view such as read also only requires the plugin to verify the caller's `SELECT` priv for the view. I don't think these behavior would change depending on the definer's privilege. Please correct me if I'm wrong.

Reply:
What I am concerned about is whether it is easy for a coder to avoid some unexpected bugs. This is related to the previous question I asked about the method `VerifyPrivilege/VerifyDynamicPrivilege`. `VerifyPrivilege` is defined as:

```
VerifyPrivilege func(activeRoles []*auth.RoleIdentity, user, host, db, table, column string, priv mysql.PrivilegeType, connState *tls.ConnectionState) bool
```

The last input argument is `connState` with type `*tls.ConnectionState`. I think you added this argument for giving different privileges for connections but with the same user, right? In this way, we can create only one user in TiDB but distinguish the "real users" from the SSL certificate. So different connections may have different privileges even they are using the same "TiDB user".

But this may bring some trouble. If a custom user "myuser1" creates a view, the definer of this view is "myuser1" by default. If another "user2" then logins without any plugin and then select this view, it should also check the privilege of "myuser1". At this time, what is the right `connState` for the method `VerifyPrivilege`? Though we can pass the tls of the current connection, but I think it is not expected. In fact, I'm not sure what is the right behavior for this scene...

Reply:
Maybe we can forbid a user to create a view/procedure if it uses custom plugin and  `VerifyPrivilege/VerifyDynamicPrivilege` is specified?

Reply:
There seems to be [two security modes](https://dev.mysql.com/doc/refman/8.0/en/create-view.html) for checking view privileges:
1. DEFINER
2. INVOKER

We can probably forbid plugin users to create view using `CREATE VIEW SQL SECURITY DEFINER ...` and only allow `CREATE VIEW SQL SECURITY INVOKER ...`, so that the `VerifyPrivilege` will only be invoked for the current connection user who is running the `SELECT` statement using its `connState`.

Reply:
@lcwangchao  I have added the above content to the design doc. Please review. Thank you!